Introduce RecipientRef interface for ActorRef and EntityRef, #24463

patriknw opened the pull request:
* to make it possible to use ActorContext.ask also with EntityRef
  destinations
* this abstraction will probably be useful for other things, where a
  full ActorRef is not required
* a MessageChannel only has to support tell (and ask), while a full ActorRef
  is watchable (destination has a lifecycle) and location transparent (serializable)

wdyt?

Refs #24463

patriknw committed:
Introduce RecipientRef interface for ActorRef and EntityRef, #24463

* to make it possible to use ActorContext.ask also with EntityRef
  destinations
* this abstraction will probably be useful for other things, where a
  full ActorRef is not required
* a RecipientRef only has to support tell (and ask), while a full ActorRef
  is watchable (destination has a lifecycle) and location transparent (serializable)

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/ActorRef.scala:
```diff
@@ -20,7 +21,7 @@ import scala.util.Success
  * (i.e. this delivery is not reliable).
  */
-trait ActorRef[-T] extends java.lang.Comparable[ActorRef[_]] with java.io.Serializable {
+trait ActorRef[-T] extends MessageChannel[T] with java.lang.Comparable[ActorRef[_]] with java.io.Serializable { this: InternalMessageChannel[T] ⇒
```
Not sure if I complicated it too much with this self type for the internal stuff. Compared with just adding the internal abstract methods here. Followed similar pattern as ActorRef. I guess the advantage is that those are completely hidden for Java.

patriknw commented on akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingSpec.scala:
```diff
@@ -288,6 +294,33 @@ class ClusterShardingSpec extends ActorTestKit with TypedAkkaSpecWithShutdown {
+          // FIXME is the implicit ClassTag difficult to use?
+          // it works fine when there is a single parameter apply,
+          // but trouble when more parameters and this doesn't compile
+          //ctx.ask(aliceRef)(x => WhoAreYou(x)) {
```
created issue https://github.com/akka/akka/issues/25611 for this

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/ActorRef.scala:
```diff
@@ -113,3 +114,27 @@ private[akka] final case class SerializedActorRef[T] private (address: String) {
+ * - not serializable
+ * - not watchable
+ */
+trait MessageChannel[-T] { this: InternalMessageChannel[T] ⇒
```
`MessageChannelRef`?

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/ActorRef.scala:
```diff
@@ -113,3 +114,27 @@ private[akka] final case class SerializedActorRef[T] private (address: String) {
+ * - not serializable
+ * - not watchable
+ */
+trait MessageChannel[-T] { this: InternalMessageChannel[T] ⇒
```
I don't like the name much, for me a channel sounds like something much more heavy weight (I think in the EIP book it is about separate programs?).

What about `Recipient` (_One that receives or is receptive._)? Or as it does not imply serialization and reference transparency `LocalRecipient`?

patriknw replied:
I like `Recipient`. We shouldn't confuse with Local. What is local? The target actor might not be local.

hepin1989 replied:
`RecipientRef` or `MessageReceiverRef`?

patriknw replied:
that's also good names, yes it might be good to add the `Ref` for this also
`ActorRef`, `EntityRef`, `SourceRef`, `SinkRef`

ktoso replied:
Agree with Johan, glad I didn't miss this PR :-) 
`RecipientRef` is excellent I think. Keeping `Ref` in the name is a good idea, all those things symbolize only a handle to the thing, not the thing itself so the Ref feels right here :)

patriknw replied:
`RecipientRef` it is then, thanks

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/InternalMessageChannel.scala:
```diff
@@ -0,0 +1,26 @@
+   */
+  def isTerminated: Boolean
+
+}
```
Could we avoid this internal extra-trait and have one that is `Aliveness` (or better name) that we only implement where it can be true and match on that in ask instead of implementing this with a constant `false` return in a bunch of places? Or maybe just put it in `MessageChannel` with a default `false` instead?

patriknw replied:
ok, I'll add such a separate internal trait, which `EntityRef` and other non `ActorRef` doesn't have to implement. Not implementing it means `false` (don't know)

ktoso reviewed (approved):
LGTM, looks good :) Cool to have this

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ActorContext.scala:
```diff
@@ -274,7 +274,7 @@ trait ActorContext[T] extends akka.actor.typed.ActorContext[T] {
   def ask[Req, Res](
     resClass:        Class[Res],
-    otherActor:      ActorRef[Req],
+    target:          MessageChannel[Req],
```
or `recipient`？

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/InternalRecipientRef.scala:
```diff
@@ -0,0 +1,26 @@
+  /**
+   * @return `true` if the actor is locally known to be terminated, `false` if alive or uncertain.
+   */
+  def isTerminated: Boolean
```
ok, I guess just keep it as is then (but I still don't like it ;) )

ktoso replied:
Like this is fine

johanandren reviewed (approved):
LGTM

patriknw commented:
Rebased and added this note to `EntityRef`:
```
   * Note that if you are inside of an actor you should prefer [[akka.actor.typed.scaladsl.ActorContext.ask]]
   * as that provides better safety.
```

Will merge after validation
